persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStore.java

```diff
@@ -282,39 +283,41 @@ private void executeLegacyUpdate(Connection connection, MarshalledEntry entry, S
    }
 
    @Override
-   public void writeBatch(Iterable<MarshalledEntry<? extends K, ? extends V>> marshalledEntries) {
+   public void writeBatch(Publisher<MarshallableEntry<? extends K, ? extends V>> publisher) {
       // If upsert is not supported, then we must execute the legacy write for each entry; i.e. read then update/insert
       if (!tableManager.isUpsertSupported()) {
-         marshalledEntries.forEach(this::write);
+         Flowable.fromPublisher(publisher).blockingForEach(this::write);
          return;
       }
 
-      Connection connection = null;
-      try {
-         connection = connectionFactory.getConnection();
-         try (PreparedStatement upsertBatch = connection.prepareStatement(tableManager.getUpsertRowSql())) {
-            int batchSize = 0;
-            for (MarshalledEntry entry : marshalledEntries) {
+      Flowable.using(() -> {
+               Connection connection = connectionFactory.getConnection();
+               PreparedStatement upsertBatch = connection.prepareStatement(tableManager.getUpsertRowSql());
+               return new KeyValuePair<>(connection, upsertBatch);
+            },
+            kvp -> createBatchFlowable(kvp.getValue(), publisher),
+            kvp -> {
+               JdbcUtil.safeClose(kvp.getValue());
+               connectionFactory.releaseConnection(kvp.getKey());
+            }
+      ).blockingSubscribe();
+   }
+
+   private Flowable<List<MarshallableEntry<? extends K, ? extends V>>> createBatchFlowable(PreparedStatement upsertBatch, Publisher<MarshallableEntry<? extends K, ? extends V>> publisher) {
+      return Flowable.fromPublisher(publisher)
+            .doOnNext(entry -> {
                String keyStr = key2Str(entry.getKey());
                prepareUpsertStatement(entry, keyStr, getSegment(entry), upsertBatch);
                upsertBatch.addBatch();
-               batchSize++;
-
-               if (batchSize == configuration.maxBatchSize()) {
-                  batchSize = 0;
-                  upsertBatch.executeBatch();
-                  upsertBatch.clearBatch();
-               }
-            }
-
-            if (batchSize != 0)
+            })
+            .buffer(configuration.maxBatchSize())
```

Comment:
So this is good. I have a possible idea of how to make it better, but not sure if it is doable with the rxjava2 API.

However I would suggest removing the first `doOnNext` and just move it into the second one and do a for loop over the list. This gives us one less lambda and should allow the JVM to optimize a bit easier.

Reply:
We could technically do a more fancy thing using scan and takeLast to avoid adding to a list. But it doesn't look that clean to me, so I would say lets just stick with the buffer we have here and change the `doOnNext`